history-network.md

```diff
@@ -0,0 +1,300 @@
+
+```
+protocol-message-type = 0x03
+protocol-message      = Container(distances: List[uint8, 256])
+```
+
+Each element of `distances` **MUST** be unique. A zero value in the `distances` list represents a request for the recipient's ENR value.
+
+#### FoundNodes (0x04)
+
+In response to a `FindNode` message, communicate a list of RLP-encoded ENR values for nodes whose `logdistance` value satisfies the request.
+
+```
+protocol-message-type = 0x04
+protocol-message      = Container(total: uint8, enrs: List[Bytes, 32])
+```
+
+The recipient may know of more than 32 nodes that satisfy the request. In this case, the recipient will need to send multiple `FoundNodes` messages in response to the `FindNode` message.
+
+Here, `total` denotes the total number of `FoundNodes` messages that the sender of the `FindNode` message should expect in response.
+
+Each ENR **MUST** be unique to some `node-id`. Each ENR **MUST** correspond to an element in the `distances` field of the `FindNode` request.
+
+#### FindContent (0x05)
+
+Request the data for `content` with `content-key`, or a list of ENR values for nodes whose `distance` from `content-id` is smaller than that of the recipient.
+
+The encoding of `content-key` for all possible data items is specified above.
+
+```
+protocol-message-type = 0x05
+protocol-message      = Container(content-key: Bytes)
+```
+
+#### FoundContent (0x06)
+
+In response to a `FindContent` message, communicate one of the following:
+
+* A connection ID for a uTP stream to transmit the requested  data
+* A list of RLP-encoded ENR values with maximum length of 32
+* A byte-array that encodes the requested data
+
+```
+protocol-message-type = 0x06
+protocol-message      = Container(connection-id: Bytes4, enrs: List[Bytes, 32], content: Bytes)
+```
+
+If the node does not hold the requested content, and the node does not know of any nodes with eligible ENR values, then the node should return `connection-id` as a zero byte array, `enrs` as an empty list and `content` as an empty byte-array.
+
+#### Offer (0x07)
+
+Communicate to the recipient that the data for `content-key` is available for transmission.
+
+```
+protocol-message-type = 0x07
+protocol-message      = Container(content-key: Bytes)
+```
+
+#### Accept (0x08)
+
+In response to a `Offer` message, request the data for `content-key`.
+
+```
+protocol-message-type = 0x08
+protocol-message      = Container(content-key: Bytes)
+```
+
+If a node transmits a `Accept` message, then we expect that node to store the corresponding data locally following the subsequent `Store` message.
+
+#### Store (0x09)
+
+In response to a `Accept` message, communicate one of the following:
+
+* A connection ID for a uTP stream to transmit the requested  data
+* A byte-array that encodes the requested data
+
+```
+protocol-message-type = 0x09
+protocol-message      = Container(connection-id: Bytes4, content: Bytes)
```

Comment:
I think the sedes here can be `Union[connection_id: Bytes4, content: Bytes]`, though the support for `Union` types in the various SSZ libraries varies (but should be simple to add).